src/Integration/Aws/SimpleS3/src/SimpleS3Client.php

```diff
@@ -21,8 +30,97 @@ public function getUrl(string $bucket, string $key): string
+    {
+        return $this->objectExists(['Bucket' => $bucket, 'Key' => $key])->isSuccess();
+    }
+
+    /**
+     * @param string|resource|callable|iterable $object
+     * @param array{
+     *   ACL?: \AsyncAws\S3\Enum\ObjectCannedACL::*,
+     *   CacheControl?: string,
+     *   ContentType?: string,
+     *   Metadata?: array<string, string>,
+     * } $options
      */
+    public function upload(string $bucket, string $key, $object, array $options = []): void
+    {
+        // Send upload requests with size 32 MB
+        $megabyte = 1024 * 1024;
+        $stream = $this->getStream($object, 32 * $megabyte);
+
+        // If file is less than 64Mb, use normal upload
+        $contentLength = $stream->length();
+        if (null !== $contentLength && $contentLength < 64 * $megabyte) {
+            $this->putObject(array_merge($options, ['Bucket' => $bucket, 'Key' => $key, 'Body' => $object]));
+
+            return;
+        }
+
+        $parts = [];
+        $partNumber = 1;
+        $uploadId = $this->createMultipartUpload(array_merge($options, ['Bucket' => $bucket, 'Key' => $key]))->getUploadId();
+        foreach ($stream as $chunk) {
+            if (1 === $partNumber && \strlen($chunk) < 5 * $megabyte) {
+                /*
+                 * The first part is too small to upload using MultipartUpload.
+                 * AWS has a limit of minimum 5MB.
+                 * Because we are using FixedSizeStream, we know that this is the
+                 * last chunk or we would had have a full "chunk size".
+                 *
+                 * Lets abort multipart upload and use a normal upload.
+                 */
+                $this->abortMultipartUpload(['Bucket' => $bucket, 'Key' => $key, 'UploadId' => $uploadId]);
+                $this->putObject(array_merge($options, ['Bucket' => $bucket, 'Key' => $key, 'Body' => $chunk]));
+
+                return;
+            }
+
+            try {
+                $response = $this->uploadPart([
+                    'Bucket' => $bucket,
+                    'Key' => $key,
+                    'UploadId' => $uploadId,
+                    'PartNumber' => $partNumber,
+                    'Body' => $chunk,
+                ]);
+                $parts[] = new CompletedPart(['ETag' => $response->getETag(), 'PartNumber' => $partNumber]);
+                ++$partNumber;
+            } catch (\Throwable $e) {
+                $this->abortMultipartUpload(['Bucket' => $bucket, 'Key' => $key, 'UploadId' => $uploadId]);
+
+                throw $e;
+            }
+        }
+
+        $this->completeMultipartUpload([
+            'Bucket' => $bucket,
+            'Key' => $key,
+            'UploadId' => $uploadId,
+            'MultipartUpload' => new CompletedMultipartUpload(['Parts' => $parts]),
+        ]);
+    }
+
+    private function getStream($object, int $chunkSize): FixedSizeStream
+    {
+        if (null === $object || \is_string($object)) {
```

Comment:
Stream factory?

Reply:
There are 2 differences between this function and the `StreamFactory`. 

1. This function sets the ChunkSize. 
1. This function returns a fixed stream

Reply:
the inner of this function can be simplified by using StreamFactory

Reply:
I dont see how if we are going to respect the `$chunkSize`